ui/src/features/Editor/components/LeftPanel/index.tsx

```diff
@@ -40,21 +41,63 @@ const LeftPanel: React.FC<Props> = ({
   onNodesAdd,
   isMainWorkflow,
   hasReader,
-  // onNodeDoubleClick,
-  // selected,
+  onNodesChange,
+  onNodeDoubleClick,
+  selected,
 }) => {
   const t = useT();
   const { workspaceId } = useParams({ strict: false });
   const [selectedTab, setSelectedTab] = useState<Tab | undefined>();
-  const { fitView } = useReactFlow();
+  const { fitView, getViewport, zoomTo } = useReactFlow();
   const [nodeId, setNodeId] = useState<string | undefined>(undefined);
 
-  useEffect(() => {
+  const previousZoomRef = useRef<number | undefined>(undefined);
+
+  const handleSelectionChange = useCallback(() => {
     if ((!isOpen && selectedTab) || (!isOpen && nodeId)) {
       setSelectedTab(undefined);
       setNodeId(undefined);
     }
-  }, [isOpen, selectedTab, nodeId]);
+
+    if (!isOpen) {
+      setNodeId(undefined);
+    }
+
+    if (selected && nodeId) {
+      if (previousZoomRef.current === undefined) {
+        previousZoomRef.current = getViewport().zoom;
+      }
+    } else if (!selected && previousZoomRef.current !== undefined) {
+      zoomTo(previousZoomRef.current, { duration: 400 });
+      previousZoomRef.current = undefined;
+    }
+  }, [isOpen, selectedTab, nodeId, selected, getViewport, zoomTo]);
+
+  useEffect(() => {
+    handleSelectionChange();
+  }, [handleSelectionChange]);
+
+  const handleTreeDataItemDoubleClick = useCallback(
+    (nodeId: string) => {
+      if (!nodeId) return;
+      const node = nodes.find((n) => n.id === nodeId);
+      if (!node) return;
+
+      fitView({
+        nodes: [{ id: node.id }],
+        duration: 500,
+        padding: 2,
+      });
+      const nodeChanges: NodeChange[] = nodes.map((n) => ({
+        id: n.id,
+        type: "select",
+        selected: n.id === nodeId,
+      }));
+      onNodesChange(nodeChanges);
+      onNodeDoubleClick({} as React.MouseEvent, node);
```

Comment:
Yeah I think making the event optional on the function and passing undefined is better than passing an empty object and type asserting.